File: scripts/import-samoa-gita.py
#!/usr/bin/env python3
"""
Import Samoa TC Gita data into PostgreSQL database
"""
import psycopg2
import json
import csv
from datetime import datetime
import sys

DATABASE_URL = "postgresql://kishank:Dcrp2024%40@localhost:5435/climate_risk"

def main():
    try:
        # Connect to database
        print("✓ Connecting to database...")
        conn = psycopg2.connect(DATABASE_URL)
        cur = conn.cursor()
        print("✓  Connected successfully")
        
        # Clear existing Gita data
        print("\n🧹 Clearing existing Samoa TC Gita data...")
        cur.execute("""
            DELETE FROM cyclone_track 
            WHERE cyclone_id = 'TC_GITA_2018' OR cyclone_name = 'Gita'
        """)
        conn.commit()
        print("✓ Cleared existing data")
        
        # Import forecast track from CSV
        print("\n📊 Importing TC Gita forecast track...")
        csv_path = '/home/kishank/Partner2/public/samoa/Official_Forecast_Track_GITA_SA.csv'
        
        imported = 0
        errors = 0
        
        with open(csv_path, 'r') as f:
            reader = csv.DictReader(f)
            for row in reader:
                try:
                    timestamp_str = row["Time[fmt=yyyy-MM-dd'T'HH:mm:ss'Z']"]
                    lat = float(row['Latitude'])
                    lon = float(row['Longitude'])
                    
                    # Normalize longitude to -180 to 180
                    if lon > 180:
                        lon = lon - 360
                    
                    pressure = float(row['Pressure']) if row['Pressure'] and row['Pressure'] != 'NaN' else None
                    wind_speed = float(row['MeanWind']) if row['MeanWind'] and row['MeanWind'] != 'NaN' else None
                    category = row.get('Category', '-3')
                    
                    # Parse timestamp
                    timestamp = datetime.strptime(timestamp_str, '%Y-%m-%d %H:%M:%S')
                    
                    # Determine if forecast (after 2018-02-10 12:00)
                    forecast_cutoff = datetime(2018, 2, 10, 12, 0, 0)
                    is_forecast = timestamp > forecast_cutoff
                    
                    # Insert into database
                    cur.execute("""
                        INSERT INTO cyclone_track (
                            cyclone_id, cyclone_name, geom, timestamp,
                            wind_speed, pressure, category, forecast
                        )
                        VALUES (
                            %s, %s, 
                            ST_SetSRID(ST_MakePoint(%s, %s), 4326),
                            %s, %s, %s, %s, %s
                        )
                    """, (
                        'TC_GITA_2018', 'Gita',
                        lon, lat,
                        timestamp, wind_speed, pressure, category, is_forecast
                    ))
                    
                    imported += 1
                    if imported % 10 == 0:
                        print(f"  Progress: {imported} points imported...", end='\r')
                        
                except Exception as e:
                    errors += 1
                    if errors < 5:
                        print(f"\n  ✗ Error at row: {e}")
        
        conn.commit()
        print(f"\n✓ Imported {imported} track points")
        if errors > 0:
            print(f"  ⚠ {errors} errors encountered")
        
        # Verify import
        print("\n✔️  Verifying import...")
        cur.execute("""
            SELECT 
                COUNT(*) as total_points,
                COUNT(*) FILTER (WHERE forecast = false) as historical_points,
                COUNT(*) FILTER (WHERE forecast = true) as forecast_points,
                MIN(timestamp) as first_point,
                MAX(timestamp) as last_point,
                MIN(pressure) as min_pressure,
                MAX(wind_speed) as max_wind_speed
            FROM cyclone_track
            WHERE cyclone_id = 'TC_GITA_2018'
        """)
        
        result = cur.fetchone()
        print(f"  Total points: {result[0]}")
        print(f"  Historical points: {result[1]}")
        print(f"  Forecast points: {result[2]}")
        print(f"  Time range: {result[3]} to {result[4]}")
        print(f"  Min pressure: {result[5]} hPa")
        print(f"  Max wind speed: {result[6]} knots")
        
        print("\n✅ Samoa TC Gita data import complete!")
        print("\n📍 You can now query the data with:")
        print("   curl http://localhost:3002/api/cyclone?cyclone_id=TC_GITA_2018")
        
        cur.close()
        conn.close()
        
    except Exception as e:
        print(f"\n❌ Import failed: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == '__main__':
    main()

File: scripts/import-samoa-gita.js
#!/usr/bin/env node
/**
 * Import Samoa TC Gita data into PostgreSQL database
 * 
 * Usage:
 *   node scripts/import-samoa-gita.js
 */

const { Client } = require('pg');
const fs = require('fs');
const path = require('path');
const { parse } = require('csv-parse/sync');

// Suppress warnings
process.removeAllListeners('warning');

// Database connection - use port 5435 for actual database
const DATABASE_URL = process.env.DATABASE_URL || 
  'postgresql://kishank:Dcrp2024%40@localhost:5435/climate_risk';

// Data file paths
const DATA_DIR = path.join(__dirname, '..', 'public', 'samoa');
const TRACK_FILE = path.join(DATA_DIR, 'cyclone-track-gita.geojson');
const FORECAST_FILE = path.join(DATA_DIR, 'Official_Forecast_Track_GITA_SA.csv');

class SamoaDataImporter {
  constructor() {
    this.client = null;
  }

  async connect() {
    this.client = new Client({ connectionString: DATABASE_URL });
    await this.client.connect();
    console.log('✓ Connected to database');
  }

  async disconnect() {
    if (this.client) {
      await this.client.end();
      console.log('✓ Disconnected from database');
    }
  }

  async clearSamoaData() {
    console.log('\n🧹 Clearing existing Samoa TC Gita data...');
    // Check if cyclone_id column exists, otherwise just clear all (assuming single cyclone)
    try {
      await this.client.query(`
        ALTER TABLE cyclone_track 
        ADD COLUMN IF NOT EXISTS cyclone_id VARCHAR(100),
        ADD COLUMN IF NOT EXISTS cyclone_name VARCHAR(255),
        ADD COLUMN IF NOT EXISTS forecast BOOLEAN DEFAULT FALSE
      `);
      await this.client.query(`
        DELETE FROM cyclone_track 
        WHERE cyclone_id = 'TC_GITA_2018' OR cyclone_name = 'Gita'
      `);
    } catch (err) {
      // If ALTER fails, just truncate the table
      await this.client.query(`TRUNCATE TABLE cyclone_track RESTART IDENTITY`);
    }
    console.log('  ✓ Cleared existing data');
  }

  async importForecastTrack() {
    console.log('\n📊 Importing Samoa TC Gita forecast track...');
    
    if (!fs.existsSync(FORECAST_FILE)) {
      console.log('  ⚠ Forecast file not found');
      return;
    }

    const fileContent = fs.readFileSync(FORECAST_FILE, 'utf-8');
    const records = parse(fileContent, {
      columns: true,
      skip_empty_lines: true,
    });

    console.log(`  Found ${records.length} track points`);

    let imported = 0;
    let errors = 0;

    for (const record of records) {
      try {
        const timestamp = record['Time[fmt=yyyy-MM-dd\'T\'HH:mm:ss\'Z\']'];
        const lat = parseFloat(record.Latitude);
        const lon = parseFloat(record.Longitude);
        
        // Normalize longitude to -180 to 180 range
        let normalizedLon = lon;
        if (lon > 180) {
          normalizedLon = lon - 360;
        }
        
        const pressure = record.Pressure && record.Pressure !== 'NaN' ? 
          parseFloat(record.Pressure) : null;
        const windSpeed = record.MeanWind && record.MeanWind !== 'NaN' ? 
          parseFloat(record.MeanWind) : null;
        const category = record.Category || '-3';
        
        // Determine if this is a forecast point (after 2018-02-10 12:00)
        const isForecast = new Date(timestamp) > new Date('2018-02-10T12:00:00Z');

        const query = `
          INSERT INTO cyclone_track (
            timestamp, 
            latitude,
            longitude,
            wind_speed, 
            pressure, 
            category,
            cyclone_id,
            cyclone_name,
            forecast
          )
          VALUES (
            $1, $2, $3, $4, $5, $6, $7, $8, $9
          )
        `;

        await this.client.query(query, [
          timestamp,
          lat,
          normalizedLon,
          windSpeed,
          pressure,
          category,
          'TC_GITA_2018',
          'Gita',
          isForecast
        ]);
        
        imported++;
        
        if (imported % 10 === 0) {
          process.stdout.write(`\r  Progress: ${imported}/${records.length}`);
        }
      } catch (error) {
        errors++;
        if (errors < 5) {
          console.error(`\n  ✗ Error importing point at ${record.Latitude},${record.Longitude}:`, error.message);
        }
      }
    }

    console.log(`\n  ✓ Imported ${imported} track points`);
    if (errors > 0) {
      console.log(`  ⚠ ${errors} errors encountered`);
    }
  }

  async verifyImport() {
    console.log('\n✔️  Verifying import...');
    
    try {
      const countResult = await this.client.query(`
        SELECT 
          COUNT(*) as total_points,
          COUNT(*) FILTER (WHERE forecast = false) as historical_points,
          COUNT(*) FILTER (WHERE forecast = true) as forecast_points,
          MIN(timestamp) as first_point,
          MAX(timestamp) as last_point,
          MIN(pressure) as min_pressure,
          MAX(wind_speed) as max_wind_speed
        FROM cyclone_track
        WHERE cyclone_id = 'TC_GITA_2018'
      `);

      const stats = countResult.rows[0];
      console.log(`  Total points: ${stats.total_points}`);
      console.log(`  Historical points: ${stats.historical_points}`);
      console.log(`  Forecast points: ${stats.forecast_points}`);
      console.log(`  Time range: ${stats.first_point} to ${stats.last_point}`);
      console.log(`  Min pressure: ${stats.min_pressure} hPa`);
      console.log(`  Max wind speed: ${stats.max_wind_speed} knots`);
    } catch (err) {
      // Fallback if cyclone_id column doesn't exist
      const countResult = await this.client.query(`
        SELECT 
          COUNT(*) as total_points,
          MIN(timestamp) as first_point,
          MAX(timestamp) as last_point,
          MIN(pressure) as min_pressure,
          MAX(wind_speed) as max_wind_speed
        FROM cyclone_track
      `);

      const stats = countResult.rows[0];
      console.log(`  Total points: ${stats.total_points}`);
      console.log(`  Time range: ${stats.first_point} to ${stats.last_point}`);
      console.log(`  Min pressure: ${stats.min_pressure} hPa`);
      console.log(`  Max wind_speed: ${stats.max_wind_speed} knots`);
    }
  }

  async run() {
    try {
      await this.connect();
      await this.clearSamoaData();
      await this.importForecastTrack();
      await this.verifyImport();
      
      console.log('\n✅ Samoa TC Gita data import complete!');
      console.log('\n📍 You can now query the data with:');
      console.log('   curl http://localhost:3002/api/cyclone?cyclone_id=TC_GITA_2018');
    } catch (error) {
      console.error('\n❌ Import failed:', error.message);
      console.error(error.stack);
      process.exit(1);
    } finally {
      await this.disconnect();
    }
  }
}

// Run the importer
const importer = new SamoaDataImporter();
importer.run();
